shell/browser/extensions/electron_extensions_browser_client.cc

```diff
@@ -127,9 +137,203 @@ base::FilePath ElectronExtensionsBrowserClient::GetBundleResourcePath(
     const base::FilePath& extension_resources_path,
     int* resource_id) const {
   *resource_id = 0;
-  return base::FilePath();
+  base::FilePath chrome_resources_path;
+  if (!base::PathService::Get(chrome::DIR_RESOURCES, &chrome_resources_path))
+    return base::FilePath();
+
+  // Since component extension resources are included in
+  // component_extension_resources.pak file in |chrome_resources_path|,
+  // calculate the extension |request_relative_path| against
+  // |chrome_resources_path|.
+  if (!chrome_resources_path.IsParent(extension_resources_path))
+    return base::FilePath();
+
+  const base::FilePath request_relative_path =
+      extensions::file_util::ExtensionURLToRelativeFilePath(request.url);
+  if (!ExtensionsBrowserClient::Get()
+           ->GetComponentExtensionResourceManager()
+           ->IsComponentExtensionResource(extension_resources_path,
+                                          request_relative_path, resource_id)) {
+    return base::FilePath();
+  }
+  DCHECK_NE(0, *resource_id);
+
+  return request_relative_path;
 }
 
+namespace {
+
+void DetermineCharset(const std::string& mime_type,
+                      const base::RefCountedMemory* data,
+                      std::string* out_charset) {
+  if (base::StartsWith(mime_type, "text/",
+                       base::CompareCase::INSENSITIVE_ASCII)) {
+    // All of our HTML files should be UTF-8 and for other resource types
+    // (like images), charset doesn't matter.
+    DCHECK(base::IsStringUTF8(base::StringPiece(
+        reinterpret_cast<const char*>(data->front()), data->size())));
+    *out_charset = "utf-8";
+  }
+}
+
+scoped_refptr<base::RefCountedMemory> GetResource(
+    int resource_id,
+    const std::string& extension_id) {
+  const ui::ResourceBundle& rb = ui::ResourceBundle::GetSharedInstance();
+  scoped_refptr<base::RefCountedMemory> bytes =
+      rb.LoadDataResourceBytes(resource_id);
+  auto* replacements =
+      ExtensionsBrowserClient::Get()->GetComponentExtensionResourceManager()
+          ? ExtensionsBrowserClient::Get()
+                ->GetComponentExtensionResourceManager()
+                ->GetTemplateReplacementsForExtension(extension_id)
+          : nullptr;
+
+  if (replacements) {
+    base::StringPiece input(reinterpret_cast<const char*>(bytes->front()),
+                            bytes->size());
+    std::string temp_str = ui::ReplaceTemplateExpressions(input, *replacements);
+    DCHECK(!temp_str.empty());
+    return base::RefCountedString::TakeString(&temp_str);
+  } else {
+    return bytes;
+  }
+}
+// Loads an extension resource in a Chrome .pak file. These are used by
+// component extensions.
+class ResourceBundleFileLoader : public network::mojom::URLLoader {
```

Comment:
Should this class be in its own file?

Reply:
updated to directly use the one from chromium.